### ISSUE TITLE
Missing Data Validation & Model Reproducibility Framework for ML Pipelines

### ISSUE THREAD

**Issue by Amogh-Koushik** [Niketkumardheeryan/ML-CaPsule#1282]
### Current Problems in ML Projects:
1. No Data Validation - Features loaded without checking for missing values, outliers, type mismatches
2. Poor Reproducibility - No fixed random seeds, inconsistent data splits, hard-coded hyperparameters
3. Incomplete Evaluation - Classification uses only accuracy (ignores precision, recall, F1), no confusion matrices
4. Inconsistent Preprocessing - Normalization, scaling, and feature engineering differ across projects
5. Missing Documentation - No explanation of data schema or preprocessing steps

Examples:
Diabetes Prediction: df = pd.read_csv('data.csv') -> no validation before training
Bird Species Classification: No evaluation metrics reported
Time Series projects: No stationarity checks before forecasting
Handwritten Recognition: Only training accuracy, no test set evaluation

### PROPOSED SOLUTION
Create a unified ML/DS framework that all data science projects can use:

3 Core Modules:

1. **Data Validation** (ml_data_validators.py)

> Check for missing values, outliers, type mismatches
> Validate class balance, feature ranges, data distribution
> Stationarity checks for time series
> 

2. **Reproducibility** (ml_reproducibility.py)

> Fixed random seeds (numpy, TensorFlow, PyTorch, Python)
> Reproducible train-test splits
> Configuration management
> Environment logging
> 

3. Model Evaluation (ml_metrics.py)

> Classification: accuracy, precision, recall, F1, ROC-AUC, confusion matrix
> Regression: RMSE, MAE, R², MAPE
> Clustering: Silhouette score, Davies-Bouldin index
> Cross-validation: K-fold with detailed stats

### What These Changes Will Do
This framework will automatically validate all incoming data for quality issues before model training begins, ensuring that missing values, outliers, and data type mismatches are caught early rather than causing silent failures during model development. It will enforce reproducible experiment execution by fixing random seeds across all libraries, guaranteeing that running the same code multiple times produces identical results, which is essential for publishing and collaborating with others. The framework will provide comprehensive model evaluation metrics tailored to different machine learning tasks—such as precision, recall, and confusion matrices for classification problems, or RMSE and R-squared for regression—replacing the current practice of reporting only accuracy or basic metrics. It will standardize data preprocessing and model training workflows across all 150+ projects by providing reusable utility modules and templates, eliminating code duplication and ensuring consistency. It will generate professional evaluation reports that clearly show model performance, identify which specific classes or predictions are problematic, and detect overfitting by comparing training versus test performance. Finally, it will establish best practices guidelines and examples that new contributors can follow, significantly lowering the barrier to entry and ensuring all future projects meet quality standards from the start.

**Desktop (please complete the following information):**
 - OS: [e.g. iOS]: WIndows
 - Browser [e.g. chrome, safari]: Chrome
 - Version [e.g. 22]: 147.0.7727.102

**Comment by parm-08** [comment]
i, I would like to work on this issue under GSSoC 2026.
I have gone through the issue details and would love to contribute.
I can have a PR ready within 3-4 days.

/assign

**Comment by shreya975** [comment]
/assign

Hi @Niketkumardheeryan,

I’d like to work on this issue as a GSSOC’26 contributor.

I reviewed the current ML pipeline structure, and introducing a proper data validation and reproducibility framework would significantly improve reliability, experiment consistency, and maintainability across training workflows.

My planned approach is:

* analyze the current preprocessing and training pipeline to identify points where inconsistent inputs or configuration drift may affect reproducibility
* implement structured data validation checks for missing values, schema consistency, feature integrity, and invalid input handling
* introduce reproducibility mechanisms such as deterministic random seeds, configuration tracking, and standardized experiment settings
* improve pipeline reliability by ensuring preprocessing and model-training stages behave consistently across executions
* add validation and regression tests for edge cases and inconsistent dataset scenarios
* keep the framework modular so future ML projects and pipelines in the repository can reuse the same validation and reproducibility utilities

I also plan to focus on maintainability and scalability so the framework can support future MLOps-oriented enhancements cleanly.

I have prior experience working on ML-focused and backend-heavy open source contributions and would love to contribute this enhancement properly.

**Comment by shreya975** [comment]
Thank you for assigning this issue to me.

**Comment by snishtha20** [comment]
Hello @niketkumardheeryan and Mentors! 

I am a GSSoC '26 contributor specializing in Python data analysis pipelines. I noticed this issue is currently unassigned and would love to work on it! 

I plan to implement a clean, beginner-friendly pipeline notebook using Pandas and NumPy to demonstrate data validation frameworks and handle missing data mismatches. Please assign this issue to me.
